program-test/src/programs.rs

```diff
@@ -17,36 +21,76 @@ mod spl_associated_token_account {
     solana_sdk::declare_id!("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
 }
 
-static SPL_PROGRAMS: &[(Pubkey, &[u8])] = &[
-    (spl_token::ID, include_bytes!("programs/spl_token-3.5.0.so")),
+static SPL_PROGRAMS: &[(Pubkey, Pubkey, &[u8])] = &[
+    (
+        spl_token::ID,
+        solana_sdk::bpf_loader::ID,
+        include_bytes!("programs/spl_token-3.5.0.so"),
+    ),
+    (
+        spl_token_2022::ID,
+        solana_sdk::bpf_loader_upgradeable::ID,
+        include_bytes!("programs/spl_token_2022-0.5.0.so"),
+    ),
     (
         spl_memo_1_0::ID,
+        solana_sdk::bpf_loader::ID,
         include_bytes!("programs/spl_memo-1.0.0.so"),
     ),
     (
         spl_memo_3_0::ID,
+        solana_sdk::bpf_loader::ID,
         include_bytes!("programs/spl_memo-3.0.0.so"),
     ),
     (
         spl_associated_token_account::ID,
+        solana_sdk::bpf_loader::ID,
         include_bytes!("programs/spl_associated_token_account-1.1.1.so"),
     ),
 ];
 
 pub fn spl_programs(rent: &Rent) -> Vec<(Pubkey, AccountSharedData)> {
     SPL_PROGRAMS
         .iter()
-        .map(|(program_id, elf)| {
-            (
+        .flat_map(|(program_id, loader_id, elf)| {
+            let mut accounts = vec![];
+            let data = if *loader_id == solana_sdk::bpf_loader_upgradeable::ID {
+                let (programdata_address, _) =
+                    Pubkey::find_program_address(&[program_id.as_ref()], loader_id);
+                let mut program_data = bincode::serialize(&UpgradeableLoaderState::ProgramData {
+                    slot: 0,
+                    upgrade_authority_address: Some(Pubkey::default()),
```

Comment:
`None` didn't work because bincode serializes them as just one zero byte, no matter the type. `Some(Pubkey::default())` ensures the proper space is used.

Reply:
Ah, ofc. I had forgotten we always have to do a separate set_authority action to set upgrade-authority to None. Thanks!